packages/ios-client/Sources/FishjamClient/pip/I420Converter.swift

```diff
@@ -0,0 +1,233 @@
+    private var poolHeight: Int = 0
+
+    init() {
+        let error = prepareForAccelerateConversion()
+
+        guard error == kvImageNoError else {
+            print("Failed to prepare for accelerate conversion: \(error)")
+            return
+        }
+    }
+
+    deinit {
+        unprepareForAccelerateConversion()
+    }
+
+    /// Creates a pixel buffer pool with specified dimensions
+    /// - Parameters:
+    ///   - width: Width of the pixel buffers
+    ///   - height: Height of the pixel buffers
+    public func createPixelBufferPool(width: Int, height: Int) {
+        poolWidth = width
+        poolHeight = height
+
+        let pixelBufferAttributes: [String: Any] = [
+            kCVPixelBufferPixelFormatTypeKey as String: kCVPixelFormatType_32BGRA,
+            kCVPixelBufferWidthKey as String: width,
+            kCVPixelBufferHeightKey as String: height,
+            kCVPixelBufferIOSurfacePropertiesKey as String: [:],
+        ]
+
+        var pool: CVPixelBufferPool?
+        let ret = CVPixelBufferPoolCreate(
+            kCFAllocatorDefault,
+            nil,
+            pixelBufferAttributes as CFDictionary,
+            &pool
+        )
+
+        if ret != kCVReturnSuccess {
+            print("Error creating pixel buffer pool: \(ret)")
+            pixelBufferPool = nil
+        } else {
+            pixelBufferPool = pool
+        }
+    }
+
+    /// Converts an I420Buffer to a CVPixelBuffer using accelerated conversion
+    /// - Parameter buffer: The I420Buffer to convert
+    /// - Returns: CVPixelBuffer or nil if conversion fails
+    public func convertToCVPixelBuffer(from buffer: RTCI420BufferProtocol) -> CVPixelBuffer? {
+        guard conversionInfo != nil else {
+            print("\(#function) failed. I420Converter failed to initialize.")
+            return nil
+        }
+
+        let width = Int(buffer.width)
+        let height = Int(buffer.height)
+
+        if pixelBufferPool == nil || poolWidth != width || poolHeight != height {
+            createPixelBufferPool(width: width, height: height)
+        }
+
+        guard let pool = pixelBufferPool else {
+            return nil
+        }
+
+        var pixelBuffer: CVPixelBuffer!
+        let status = CVPixelBufferPoolCreatePixelBuffer(kCFAllocatorDefault, pool, &pixelBuffer)
+
+        guard status == kCVReturnSuccess else {
+            return nil
+        }
+
+        let error = convertFrameVImageYUV(buffer, toBuffer: pixelBuffer)
+
+        if error != kvImageNoError {
+            print("\(#function): error during conversion: \(error)")
+            return nil
+        }
+
+        return pixelBuffer
+    }
+
+    // MARK: - Private Methods
+
+    /// Prepares the converter for accelerated conversion using vImage
+    /// - Returns: vImage_Error indicating success or failure
+    private func prepareForAccelerateConversion() -> vImage_Error {
+        // Setup the YpCbCr to ARGB conversion.
+        if conversionInfo != nil {
```

Comment:
Here again we are running into those potential race conditions with the `conversionInfo`. I think it would be good to figure out the strategy we want to follow in the project to avoid them.

Reply:
I'm not sure what errors are you taking about? 🤔 
This code is mostly just obj-c => swift conversion takes from here: https://github.com/twilio/video-ios-affectiva/blob/ed2e864324c40ad25e5a06cc2b05298b03caed09/EmoCall/I420Converter.m

Reply:
But when I look at the code now. `prepareForAccelerateConversion` is only called from the init, so this check will never be called anyway.